dictionaries/CallMap_80_delta.php

```diff
@@ -941,6 +941,10 @@
       'old' => ['bool', 'mysql'=>'mysqli', 'flags='=>'int', 'name='=>'string'],
       'new' => ['bool', 'mysql'=>'mysqli', 'flags='=>'int', 'name='=>'?string'],
     ],
+    'openssl_open' => [
+      'old' => ['bool', 'data'=>'string', '&w_output'=>'string', 'encrypted_key'=>'string', 'private_key'=>'string|array|resource', 'cipher_algo='=>'string', 'iv='=>'string'],
+      'new' => ['bool', 'data'=>'string', '&w_output'=>'string', 'encrypted_key'=>'string', 'private_key'=>'OpenSSLAsymmetricKey|OpenSSLCertificate|array|string', 'cipher_algo='=>'string', 'iv='=>'string'],
```

Comment:
Doesn't `$private_key` here have the same range of types as in `openssl_sign`? That one has more detailed type for its `array` union element.

Reply:
I took this signature from the [php manual](https://www.php.net/openssl_open), but yes, you're right. I overlooked that psalm of course has more info about arrays. let me fix that.

Reply:
oh but wait: Does that signature make sense?

it's currently `array{OpenSSLAsymmetricKey|OpenSSLCertificate|string, string}` what does that even mean in psalm? using {} implies an object-like array, but then shouldn't there be a : key/value separator?

also: Where did we take the shape of that array from? The [php manual](https://www.php.net/openssl_sign) doesn't actually explain the array case

Reply:
and looking at the php source code, I think, this is definitely bogus and should be

`array{0: string, 1: string}`

https://github.com/php/php-src/blob/35786e321b64cb064d9bbdc339836b7b9d1d1800/ext/openssl/openssl.c#L3509

Reply:
I also just noticed that while `openssl_pkey_get_private` got the PHP-8 treatment, its alias `openssl_get_privatekey` didn't. and neither did `openssl_pkey_get_public` nor its alias `openssl_get_publickey`.

I'll fix all of those too, but can you first confirm that the currently present definition for the `array` argument is suboptimal? Then I'll fix all of those too.

Reply:
`array{A, B}` is a shorthand for `array{0:A, 1:B}`.

> Where did we take the shape of that array from?

There's a separate chapter on key / certificate params that vaguely documents array shape used: https://www.php.net/manual/en/openssl.certparams.php

@jirkace experimented with objects as the first key element and found it worked: https://github.com/vimeo/psalm/pull/6846#issuecomment-962867211

Reply:
> array{A, B} is a shorthand for array{0:A, 1:B}

I believe there is a slight difference
`[1 => 'B', 0 => 'A']` would match the second one but not the first because the first one imply the order of the keys is important

Inside of Psalm, this flag is know as `is_list`, which suggest it has to start from 0 and be in order

Reply:
> I'm not sure this is intended from the PHP side.

IMO that would make sense - you have to provide passphrase somehow.

Reply:
But in order to get to that `OpenSSLAsymmetricKey ` instance to pass in the array, you need to already have provided the passphrase to `openssl_pkey_get_private `

Reply:
Perhaps. I'm only superficially acquainted with this extension so I'm inclined to take your word for it.